docs/docs/aztec/writing_efficient_contracts.md

```diff
@@ -0,0 +1,62 @@
+---
+title: Writing Efficient Contracts
+sidebar_position: 2
+tags: [Developers, Contracts]
+---
+
+## Writing functions
+
+On Ethereum L1, all data is public and all execution is completely reproducible. The Aztec L2 takes on the challenge of execution of private functions on private data. This is done client side, along with the generation of corresponding proofs, so that the network can verify the proofs and append any encrypted data/nullifiers (privacy preserving state update).
+
+This highlights a key difference with how public vs private functions are written.
+
+:::info Writing efficiently
+
+- **Public functions** can be written intuitively - optimising for execution/gas as one would for EVM L2s
+- **Private functions** are optimised differently, as they are compiled to a circuit to be proven (see [Thinking in Circuits](https://noir-lang.org/docs/explainers/explainer-writing-noir))
+
+:::
+
+## Assessing efficiency
+
+On Aztec (like other L2s) there are several costs/limit to consider...
+
+- L1 costs - execution, blobs, events
+- L2 costs - public execution, data, logs
+- Local limits - proof generation, execution
+
+### Local Proof generation
```

Comment:
I think it would be nice to include expected gate counts for common operations, like in rahul's doc

e.g.
~75 gates per poseidon2 hash of 3 fields 
3500 for reading on a tree (e.g. nullifier tree, note hash tree or public data tree)
4000 for a shared mutable read
X000 for sha256
X000 to constrained-encrypt a log of Y fields
X000 to constrained-encrypt-and-tag a log of Y fields

Reply:
See comment [here](https://github.com/AztecProtocol/aztec-packages/pull/14734#discussion_r2126521062).